### PR TITLE
Domain Sockets

### PR DESCRIPTION
**Pull request by rennergade**
## Description 
Configuring domain sockets to use Lind's user space pipes

joint PR from work by @justyoyo123 @TristanB22 @yizhuoliang @RusherRG and @rennergade 

Fixes # [(issue)](https://github.com/Lind-Project/lind_project/issues/186)

This changes domain sockets to use a set of lind pipe internally in user space as well as refactors networking calls to clearly demonstrate each type of socket

### Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Added tests in ipc_tests

Additional tests on lind_project branch

Working postgres/LAMP/pgbench

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/net_calls.rs [hunk 1]
```diff
@@ -1034,162 +1232,216 @@ impl Cage {
...
+       };
  
-        let mut retval = 0; 
-        loop { //we must block manually
-            for fd in readfds.iter() {
+       let mut retval = 0; 
+       loop { //we must block manually
+           
+            // 1. iterate thru readfds
+            let mut res = self.select_readfds(readfds, &mut new_readfds, &mut retval);
```
Comment (justyoyo123) [inline src/safeposix/syscalls/net_calls.rs:1258]: Just curious, I forget why this change with select was needed. Was it due to read/write locks causing issues?

Comment (rennergade) [reply src/safeposix/syscalls/net_calls.rs:1258]: Are you asking why we refactored to separate the read and write parts of select? Mostly for readability.

File: src/safeposix/syscalls/net_calls.rs [hunk 2]
```diff
@@ -53,12 +52,6 @@ impl Cage {
...
 
-    pub fn swap_unixaddr(remoteaddr: &interface::GenSockaddr) -> interface::GenSockaddr {
-        let mut swapaddr = remoteaddr.clone();
-        if let Some(addr) = NET_METADATA.revds_table.get(&remoteaddr.clone()) { swapaddr = *addr; }
-        swapaddr
-    }
-
     pub fn socket_syscall(&self, domain: i32, socktype: i32, protocol: i32) -> i32 {
         let real_socktype = socktype & 0x7; //get the type without the extra flags, it's stored in the last 3 bits
         let nonblocking = (socktype & SOCK_NONBLOCK) != 0;
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs:57]: @yizhuoliang can we remove this blocking error and also clean up the language so the blocking/nonblocking are the same in here and the socket initializer

File: src/safeposix/syscalls/net_calls.rs [hunk 3]
```diff
@@ -131,112 +129,101 @@ impl Cage {
...
+            AF_INET | AF_INET6 => self.bind_inner_socket_inet(sockhandle, &mut newsockaddr, prereserved),
+            _ => {return syscall_error(Errno::EINVAL, "bind", "Unsupported domain provided");}
+        };
 
-        if sockhandle.realdomain == AF_UNIX {
-            // create fake IPV4 addr
-            let ipaddr = interface::V4Addr {s_addr: u32::from_ne_bytes([127, 0, 0, 1])};
-            let innersockaddr = interface::SockaddrV4{sin_family: AF_INET as u16, sin_addr: ipaddr, sin_port: 0, padding: 0};
-            newsockaddr = interface::GenSockaddr::V4(innersockaddr);
+        if res != 0 {
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs:148]: I think we can remove this res check no with the refactor @yizhuoliang

Comment (yizhuoliang) [reply src/safeposix/syscalls/net_calls.rs]: yes, removed

File: src/safeposix/syscalls/net_calls.rs (hunk 2 above)
Comment (yizhuoliang) [reply src/safeposix/syscalls/net_calls.rs:57]: done

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -1145,7 +1141,7 @@ impl Cage {
             if fdguard.is_some() {
                 drop(fdguard);
                 //close the fd in the way of the new fd. If an error is returned from the helper, return the error, else continue to end
-                let close_result = Self::_close_helper_inner(&self, newfd);
+                let _close_result = Self::_close_helper_inner(&self, newfd);
                 // mirror the implementation of linux, ignore the potential error of the close here
             } else { drop(fdguard); }
```
Comment (kuzeyardabulut) [inline src/safeposix/syscalls/fs_calls.rs:1146]: Can we change that part like below:

```
 drop(fdguard);
 
 if !fdguard.is_some() {
      let _close_result = Self::_close_helper_inner(&self, newfd);
 }
```

I think this one can be more legible.

File: src/safeposix/syscalls/fs_calls.rs [hunk 5]
```diff
@@ -339,10 +339,10 @@ impl Cage {
...
                 let mut inodeobj = FS_METADATA.inodetable.get_mut(&inodenum).unwrap();
 
-                let (currefcount, curlinkcount, has_fobj, has_domsock) = match *inodeobj {
-                    Inode::File(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, true, false)},
-                    Inode::CharDev(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, false, false)},
-                    Inode::Socket(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, false, true)},
+                let (currefcount, curlinkcount, has_fobj, log) = match *inodeobj {
+                    Inode::File(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, true, true)},
+                    Inode::CharDev(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, false, true)},
+                    Inode::Socket(ref mut f) => {f.linkcount -= 1; (f.refcount, f.linkcount, false, false)},
```
Comment (TristanB22) [inline src/safeposix/syscalls/fs_calls.rs:345]: Is there a reason for suddenly inverting these boolean values?

File: src/safeposix/syscalls/net_calls.rs [hunk 6]
```diff
@@ -551,148 +619,177 @@ impl Cage {
...
 
+        let bufleft = newbufptr;
+        let buflenleft = newbuflen;
+        let mut retval;
+
+        if sockhandle.domain == AF_UNIX {
+             // get the remote socket pipe, read from it, and return bytes read
+            let mut nonblocking = false;
+            if sockfdobj.flags & O_NONBLOCK != 0 { nonblocking = true;}  
+            loop {
```
Comment (TristanB22) [inline src/safeposix/syscalls/net_calls.rs:678]: can we add justification for looping here in a comment?

File: src/safeposix/syscalls/net_calls.rs [hunk 7]
```diff
@@ -551,148 +619,177 @@ impl Cage {
...
+                            //the case that this recv timeout was exceeded, and we
+                            //should thus not treat this as a failure in our emulated
+                            //socket; see comment in Socket::new in interface/comm.rs
+                            if sockfdobj.flags & O_NONBLOCK == 0 && i == Errno::EAGAIN {
+                                if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
+                                    // if the cancel status is set in the cage, we trap around a cancel point
+                                    // until the individual thread is signaled to cancel itself
+                                    loop { interface::cancelpoint(self.cageid); }
                                 }
+                                interface::RustLockWriteGuard::<SocketHandle>::bump(sockhandle);
```
Comment (TristanB22) [inline src/safeposix/syscalls/net_calls.rs:731]: double check this for potential deadlock if we keep getting EAGAIN

File: src/tests/ipc_tests.rs [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    pub fn test_ipc() {
+        ut_lind_ipc_pipe();
+        ut_lind_ipc_domain_socket();
+        ut_lind_ipc_socketpair(); 
+    }
+
+    pub fn ut_lind_ipc_pipe() {
+
+        let byte_chunk: usize = 131072;
+        let num_writes: usize = 8192;
```
Comment (TristanB22) [inline src/tests/ipc_tests.rs:20]: Can we change these and the other values to their 2<<x representations? like 2<<17 and 2<<13

**Review by TristanB22 — APPROVED** [review]
Overall looks fine. I didn't see any glaring logic errors and wouldn't be able to spot the smaller ones. Just some legibility comments from my end.

File: src/tests/networking_tests.rs [hunk 9]
```diff
@@ -758,13 +768,20 @@ pub mod net_tests {
...
                 } else { //If the socket is in established conn., then we recv the data. If there's no data, then close the client socket.
                     let mut buf = sizecbuf(4);
-                    let recvresult = cage.recv_syscall(sock, buf.as_mut_ptr(), 4, 0);
+                    let mut recvresult :i32;
+                    loop {
+                        recvresult = cage.recv_syscall(sock, buf.as_mut_ptr(), 4, 0);
+                        if recvresult == -libc::EINTR {
+                            continue; // if the error was EINTR, retry the syscall
+                        }
+                        break;
```
Comment (kuzeyardabulut) [inline src/tests/networking_tests.rs:783]:
```suggestion
                        if recvresult != -libc::EINTR {
                            break; // if the error was EINTR, retry the syscall
                        }
```

File: src/tests/networking_tests.rs [hunk 10]
```diff
@@ -965,70 +980,52 @@ pub mod net_tests {
...
         let mut buf2 = sizecbuf(15);
-        cage.recv_syscall(socketpair.sock1, buf2.as_mut_ptr(), 15, 0);
-        assert_eq!(cbuf2str(&buf2), "Socketpair Test");
-    
+        loop {
+            let result = cage.recv_syscall(socketpair.sock1, buf2.as_mut_ptr(), 15, 0);
+            if result == -libc::EINTR {
+                continue; // if the error was EINTR, retry the syscall
+            }
+            break;
```
Comment (kuzeyardabulut) [inline src/tests/networking_tests.rs:1013]:
```suggestion
            if result != -libc::EINTR {
                break; // if the error was EINTR, retry the syscall
            }
```

File: src/interface/comm.rs [hunk 11]
```diff
@@ -166,7 +166,7 @@ pub fn new_sockaddr_unix(family: u16, path: &[u8]) -> SockaddrUnix {
 }
 
 pub fn gen_ud_path() -> String {
-    let mut owned_path: String = "/tmp/sock".to_owned();
+    let mut owned_path: String = "/sock".to_owned();
```
Comment (rennergade) [inline src/interface/comm.rs:169]: I forget why we changed this to just /sock from /tmp/sock.

It would make sense to go in /tmp, we should check this again to see if that breaks anything.

File: src/safeposix/syscalls/fs_calls.rs (hunk 5 above)
Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs:345]: they are different things actually, before we had to manage domain sockets separately which the has_domsock variable checked for, the log variable is for the log based file system

File: src/safeposix/syscalls/fs_calls.rs (hunk 4 above)
Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs:1146]: Your replacement doesn't exactly work because we need to check is_some before the fdguard is dropped. I re-factored a bit to make it cleaner.

File: src/safeposix/syscalls/net_calls.rs (hunk 7 above)
Comment (rennergade) [reply src/safeposix/syscalls/net_calls.rs:731]: its a blocking socket so we should loop continuously. were just hijacking EAGAIN here to check for cancellation

File: src/safeposix/net.rs [hunk 12]
```diff
@@ -110,16 +108,65 @@ impl Drop for SocketHandle {
...
     next_ephemeral_port_udpv4: interface::RustRfc<interface::RustLock<u16>>,
     next_ephemeral_port_tcpv6: interface::RustRfc<interface::RustLock<u16>>,
     next_ephemeral_port_udpv6: interface::RustRfc<interface::RustLock<u16>>,
     pub listening_port_set: interface::RustHashSet<(interface::GenIpaddr, u16, PortType)>,
-    pub domain_socket_table: interface::RustHashMap<interface::RustPathBuf, interface::GenSockaddr>,
-    pub revds_table: interface::RustHashMap<interface::GenSockaddr, interface::GenSockaddr>,
-    pub pending_conn_table: interface::RustHashMap<u16, Vec<(Result<interface::Socket, i32>, interface::GenSockaddr)>>
+    pub pending_conn_table: interface::RustHashMap<u16, Vec<(Result<interface::Socket, i32>, interface::GenSockaddr)>>,
+    pub domsock_accept_table: interface::RustHashMap<interface::RustPathBuf, DomsockTableEntry>,
+    pub domsock_paths: interface::RustHashSet<interface::RustPathBuf>
```
[the remark this answers is not in the document]
Comment (rennergade) [reply src/safeposix/net.rs:169]: they are different but it probably needs clarification.

The paths set is the set of all domain socket paths that have been bound, where the accept_table manages UDS connections/accepts. Fundamentally they do different things.

[the remark this answers is not in the document]
Comment (rennergade) [reply src/safeposix/net.rs:169]: good question though

**Comment by Yaxuan-w** [comment]
Fantastic work! The modifications not only make the project compatible with the latest version of Rust but also enhance its adaptability to a broader range of tools. But it generates 8 warning with `rustc 1.75.0-nightly`:
Example warnings:
```
warning: call to `.clone()` on a reference in this situation does nothing
  --> src/interface/comm.rs:74:57
   |
74 |                 let pathslice = pathiter.next().unwrap().clone();
   |                                                         ^^^^^^^^ help: remove this redundant call
   |
   = note: the type `[u8]` does not implement `Clone`, so calling `clone` on `&[u8]` copies the reference, which does not do anything and can be removed
   = note: `#[warn(noop_method_call)]` on by default
```
